### PR TITLE
Reduce datapath from_lxc complexity

### PR DESCRIPTION
**Pull request by jrajahalme**
Split per packet load balancing into a separate tail call. This reduces datapath complexity when per packet load balancing is needed.

Preparatory commits simplify tail call logic and refactors IPv6 from LXC code paths to be more similar to IPv4. This makes edits and reviews on the functional change commits easier.

This PR should have minimal effect on the code when per packet load balancing is not required. On IPv4 there should be no change, and on IPv6 there is one extra data revalidation due to logic being split into two tail callable functions.

This is a prerequisite for #18894

### CODE REVIEW

**Comment by jrajahalme** [comment]
Added `policy_clear_mark()` call to the per packet lb path as the new `CB_BACKEND_ID` aliases with `CB_POLICY`.

**Comment by joestringer** [comment]
> This may be due to the new tail calls I added, so the map size changed. Is there something I could do to allow upgrade/downgrade to work? In general I think we should be able to add new tail calls in new versions, no?

Unfortunately not, we don't have a way to safely migrate tail call maps. I suspect that https://github.com/cilium/cilium/pull/13032 is intended to allow this, but that PR has been stalled for some time.

**Comment by jrajahalme** [comment]
I was told #17744 may fix this error on prog map size change. Can we please merge it :-)

File: bpf/bpf_lxc.c
Comment (kkourt) [inline bpf/bpf_lxc.c]: Could it be that the difference here was intentional due to the different sizes between v4 and v6 tuples?

File: bpf/bpf_lxc.c
Comment (kkourt) [inline bpf/bpf_lxc.c]: FWIW, I believe the original intention of this was to maintain the code structure of the previous version when we checked for an OR instead of an AND. But, maybe it does not make a lot of sense to keep it.

**Review by kkourt — COMMENTED** [review]
Just some minor comments from a first cursory pass :)

File: bpf/lib/policy.h
Comment (pchaigno) [inline bpf/lib/policy.h:324]: That doesn't seem intended. @aanm Is that already on your radar?

File: bpf/bpf_lxc.c
Comment (pchaigno) [inline bpf/bpf_lxc.c:106]: I'm surprised checkpatch doesn't complain about the line being too long here :thinking:

File: bpf/bpf_lxc.c
Comment (pchaigno) [reply bpf/bpf_lxc.c]: I think this is indeed intentional to rely on [our own `memcpy` variant](https://github.com/cilium/cilium/blob/4f27865e975cc095891b68df5f4a7bc0e26ecca5/bpf/include/bpf/builtins.h#L154), which can optimize the number of instructions necessary for the copy.

File: bpf/bpf_lxc.c
Comment (pchaigno) [inline bpf/bpf_lxc.c:454]: Nice! That has been annoying me for so long...

File: bpf/bpf_lxc.c
Comment (pchaigno) [reply bpf/bpf_lxc.c]: No no, I just didn't really think about what I was doing :smile:

File: bpf/bpf_lxc.c
Comment (pchaigno) [inline bpf/bpf_lxc.c]: Commit `datapath: Reduce from LXC complexity` says:

> IPv6 code path is refactored to pattern it more similar to IPv4 code path.

That's not the case anymore, right? The refactoring was moved to other commits?

File: bpf/bpf_lxc.c
Comment (pchaigno) [inline bpf/bpf_lxc.c]: The logic to encode/decode the values from slot `CB_CT_STATE` could be moved to helper functions. Especially given we have the exact same logic on the IPv4 side.

**Review by pchaigno — APPROVED** [review]
Thanks! Only nits below :point_down: 

(Oh, and thanks a lot for the clean commit history. Super easy to review! :pray:)

File: bpf/lib/policy.h
Comment (aanm) [reply bpf/lib/policy.h:324]: What's not intended?

File: bpf/bpf_lxc.c
Comment (jrajahalme) [reply bpf/bpf_lxc.c]: OK, reverted back.

File: bpf/lib/policy.h
Comment (jrajahalme) [reply bpf/lib/policy.h:324]: @aanm Maybe this updated comment can be revised later if the behavior is changed?

File: bpf/bpf_lxc.c
Comment (jrajahalme) [reply bpf/bpf_lxc.c:106]: I recall it complains of lines longer than 100 characters?

File: bpf/bpf_lxc.c
Comment (jrajahalme) [reply bpf/bpf_lxc.c]: Thanks, removed this sentence from the commit message :-)

File: bpf/lib/policy.h
Comment (pchaigno) [reply bpf/lib/policy.h:324]: @jrajahalme Yes, definitely not blocking this PR. Thanks for fixing the comment!

File: bpf/bpf_lxc.c
Comment (qmonnet) [reply bpf/bpf_lxc.c:106]: Jarno is correct https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?h=bdc48fa11e46f867ea4d75fa59ee87a7f48be144

File: bpf/bpf_lxc.c
Comment (jrajahalme) [reply bpf/bpf_lxc.c]: Done!

**Review by kkourt — APPROVED** [review]
LGTM, thanks!

**Review by pchaigno — APPROVED** [review]

**Comment by jrajahalme** [comment]
rebased and added setting `tuple.nexthdr` that I missed before: https://github.com/cilium/cilium/blob/3530b7aa820de66aea2d755df282952cdeb01038/bpf/bpf_lxc.c#L485

**Comment by jrajahalme** [comment]
@pchaigno Do you think #17744 being in limbo should keep blocking this PR?

**Comment by pchaigno** [comment]
> @pchaigno Do you think #17744 being in limbo should keep blocking this PR?

Without #17744, the upgrade test will fail, no? I don't see how we can merge this safely without #17744 :disappointed:

**Comment by jrajahalme** [comment]
@pchaigno Rebased to address a conflict and added a new commit to always set the state instead of depending on the datapath having cleared skb CB to begin with, and clearing also `CB_CT_STATE` after restore to not leak it to possible later tail calls.

I'll squash the new commit in to the main one if you think this is an acceptable change :-)

**Comment by jrajahalme** [comment]
test-runtime

**Comment by jrajahalme** [comment]
runtime test failed to provision a VM, retrying.

**Comment by jrajahalme** [comment]
EKS test hit the maximum number of VPCs, failed to create a cluster because of it.

**Comment by pchaigno** [comment]
Switching to draft mode to remove it from tophat/review lists. Hopefully it won't be too long before we can merge :crossed_fingers:

**Comment by jrajahalme** [comment]
Still blocked on #17744

**Comment by jrajahalme** [comment]
#17744 merged, now this is ready to review :-)

File: bpf/lib/policy.h
Comment (pchaigno) [reply bpf/lib/policy.h:324]: I've opened https://github.com/cilium/cilium/issues/18811 for this.

File: bpf/lib/common.h
Comment (pchaigno) [inline bpf/lib/common.h]: Nit: This is using spaces instead of tabs between CB_POLICY and the comment.

File: bpf/bpf_lxc.c
Comment (pchaigno) [inline bpf/bpf_lxc.c]: Shouldn't commit `datapath: Be more robust about state between tail calls` be merged with the previous commit? As far as I can see, it only touches lines introduces by the previous commit.

**Review by pchaigno — APPROVED** [review]

File: bpf/bpf_lxc.c
Comment (jrajahalme) [reply bpf/bpf_lxc.c]: Good point, squashed :-)

**Comment by jrajahalme** [comment]
test-1.16-4.9

**Comment by jrajahalme** [comment]
Upgrade/downgrade tests will keep on failing until 1.11 is released with recently merged backport (https://github.com/cilium/cilium/pull/18800)

**Comment by jrajahalme** [comment]
rebased

**Comment by jrajahalme** [comment]
runtime test provisioning had failed, restarted

**Comment by jrajahalme** [comment]
@pchaigno upgrade/downgrade test is still failing, any ideas why?

**Comment by jrajahalme** [comment]
@joestringer, @pchaigno This is blocking https://github.com/cilium/cilium/pull/18894, do we have any idea of what would be needed to fix the upgrade/downgrade in the presence of changes in the calls map between the versions? My understanding so far is that upgrade "works" but some traffic will be impacted due to mismatching bpf programs w.r.t. calls map?

**Comment by ti-mo** [comment]
I'm looking into it.

**Comment by ti-mo** [comment]
Backport pending in https://github.com/cilium/cilium/pull/19066.

**Comment by jrajahalme** [comment]
#19066 was just merged, retesting...

**Comment by jrajahalme** [comment]
ci-gke test failed due to curl apparently returning exit code 0 even when the traffic was dropped (https://github.com/cilium/cilium-cli/issues/757). Testing again to see if this was a flake.

**Comment by jrajahalme** [comment]
Verifier test failed:
```
Failed to load BPF program bpf_lxc with datapath configuration:
-DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 -DENABLE_HOST_SERVICES_TCP=1 -DENABLE_HOST_SERVICES_UDP=1 -DENABLE_HOST_REDIRECT=1 -DENABLE_ROUTING=1 -DNO_REDIRECT=1 -DPOLICY_VERDICT_NOTIFY=1 -DALLOW_ICMP_FRAG_NEEDED=1 -DENABLE_IDENTITY_MARK=1 -DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1 -DHAVE_LPM_TRIE_MAP_TYPE=1 -DHAVE_LRU_HASH_MAP_TYPE=1 -DENABLE_MASQUERADE=1 -DENABLE_SRC_RANGE_CHECK=1 -DENABLE_NODEPORT=1 -DENABLE_NODEPORT_ACCELERATION=1 -DENABLE_SESSION_AFFINITY=1 -DENABLE_DSR_ICMP_ERRORS=1 -DENABLE_DSR=1 -DENABLE_DSR_HYBRID=1 -DENABLE_IPV4_FRAGMENTS=1 -DENCAP_IFINDEX=1 -DTUNNEL_MODE=1
Expected command: kubectl exec -n default test-verifier -- env TC_PROGS="" XDP_PROGS="" CG_PROGS="" TC_PROGS="bpf_lxc" ./test/bpf/verifier-test.sh 
To succeed, but it failed:
Exitcode: 1 
Err: exit status 1
Stdout:
 	 => Loading bpf_lxc.c:from-container...
	 
Stderr:
```
...
```
	 BPF program is too large. Processed 131073 insn
	 
	 libbpf: -- END LOG --
	 libbpf: failed to load program 'tail_handle_ipv4'
```

File: bpf/bpf_lxc.c
Comment (jrajahalme) [inline bpf/bpf_lxc.c:55]: This change was needed for `declare_tailcall_if(is_defined(ENABLE_PER_PACKET_LB), ...` to work as expected. Without this change the tailcall got never declared and the declared function was called inline, causing a verifier failure due to too large program.

Comment (ti-mo) [reply bpf/bpf_lxc.c:55]: Hmmm, that sounds like a red herring. Tail calls are never directly invoked and thus cannot be inlined. Are you sure?

Comment (pchaigno) [reply bpf/bpf_lxc.c:55]: No no, that makes sense. `is_defined` expects a value for the macros. We've had this issue in the past.
Then if it returns false, `declare_tailcall_if` wouldn't compile into a tail call.

> Tail calls are never directly invoked and thus cannot be inlined.

Not sure what you mean by this.

**Comment by jrajahalme** [comment]
Runtime provisioning fail, testing.

**Comment by jrajahalme** [comment]
Cilium pre-flight checks timing out without any agent errors/warnings: https://jenkins.cilium.io/job/Cilium-PR-K8s-1.21-kernel-5.4/732/. Retesting to see if this was a flake.

**Comment by jrajahalme** [comment]
/test-1.21-5.4

File: bpf/bpf_lxc.c
Comment (ti-mo) [reply bpf/bpf_lxc.c:55]: Re: inlining: I see, inlining vs. emitting functions as tail calls seems to be the whole point of `declare_tailcall_...`. :slightly_smiling_face: Had never explored the declaration.

**Comment by jrajahalme** [comment]
runtime provisioning fail, retesting

**Comment by jrajahalme** [comment]
BPF complexity issue on kernel 4.9:
```
2022-03-18T13:52:39.660513642Z level=warning msg="libbpf: failed to load program 'tail_handle_ipv6_cont'" subsys=datapath-loader
2022-03-18T13:52:39.660515991Z level=warning msg="libbpf: failed to load object '3064_next/bpf_lxc.o'" subsys=datapath-loader
```

**Comment by michi-covalent** [comment]
this one looks like a legitimate failure https://jenkins.cilium.io/job/Cilium-PR-K8s-1.16-kernel-4.9/1074/ 👀

**Comment by jrajahalme** [comment]
Fixed the verifier/complexity issue in local testing, will squash the last 2 commits if the CI agrees.

**Comment by jrajahalme** [comment]
Still failing test-verifier test on kernel 4.9, but no verifier errors in any end-to-end CI tests.

**Comment by jrajahalme** [comment]
> Still failing test-verifier test on kernel 4.9, but no verifier errors in any end-to-end CI tests.

@pchaigno The test does not fail if I don't define `-DSKIP_DEBUG=1`...

**Comment by jrajahalme** [comment]
Opened a new issue for cilium-cli flake for ci-gke: https://github.com/cilium/cilium-cli/issues/763

**Comment by jrajahalme** [comment]
5.4 known flake: #16122

**Comment by jrajahalme** [comment]
/test-1.21-5.4

**Comment by jrajahalme** [comment]
CI was all green, but had to squash in the fixes in the last commit.

**Comment by jrajahalme** [comment]
Triggered tests too early, they timed out waiting for images..

**Comment by jrajahalme** [comment]
ci-gke hit by known flake #19122

**Comment by jrajahalme** [comment]
Runtime test hit by known flake #19124

File: bpf/bpf_lxc.c
Comment (pchaigno) [inline bpf/bpf_lxc.c:144]: I'm guessing the reason we were making that lookup after the CT lookup is to avoid one lookup in case we need to send the packet to the proxy. Why did the verifier want this here? Was it a complexity issue or some other error?

File: bpf/bpf_lxc.c
Comment (pchaigno) [inline bpf/bpf_lxc.c:168]: What was the verifier error here? Was it a complexity issue?

File: bpf/bpf_lxc.c
Comment (pchaigno) [inline bpf/bpf_lxc.c:334]: What's the reason for this change?

**Review by pchaigno — CHANGES_REQUESTED** [review]
I'm a bit worried about moving things around because there was a verifier error without really knowing why it complained. Unless it was a complexity issue, there's probably a better fix.

The rest looks good to me.

File: bpf/bpf_lxc.c
Comment (jrajahalme) [reply bpf/bpf_lxc.c:144]: Complexity issue in kernel 4.9.

File: bpf/bpf_lxc.c
Comment (jrajahalme) [reply bpf/bpf_lxc.c:168]: Yes, complexity issue in the test-verifier on kernel 4.9 after the complexity issue on the end-to-end test was fixed by reshuffling the code.

File: bpf/bpf_lxc.c
Comment (jrajahalme) [reply bpf/bpf_lxc.c:334]: Part of limiting the scope of variables in the stack to try resolve the complexity issue on kernel 4.9.

**Comment by jrajahalme** [comment]
> I'm a bit worried about moving things around because there was a verifier error without really knowing why it complained. Unless it was a complexity issue, there's probably a better fix.
> 
> The rest looks good to me.

@pchaigno All the code movement was to address verifier complexity issues on kernel 4.9. I already spent a full weekend getting all this passing. If you think some of these changes should be rolled back, I would appreciate if you could figure out which changes can be rolled back without hitting 4.9 complexity issues again :-)

**Review by brb — APPROVED** [review]
:rocket:

File: bpf/bpf_lxc.c
Comment (pchaigno) [reply bpf/bpf_lxc.c:334]: That shouldn't affect complexity. It shouldn't even affect the generated bytecode as far as I can see.

File: bpf/bpf_lxc.c
Comment (pchaigno) [inline bpf/bpf_lxc.c:170]: Ok, so after some testing with this branch and https://github.com/jrajahalme/cilium/pull/new/wip-datapath-reduce-from-lxc-complexity, it turns out the complexity issue isn't "fixed" here because of the helper call, but simply because LLVM reshuffles code around once you add `revalidate_data`. That's why adding `relax_verifier` instead doesn't help.

So this is very fragile and likely to still cause complexity issues elsewhere. I don't have a better solution though. Just note that we don't have complete coverage for complexity issues so we might be adding more here than we are fixing :grimacing:

**Review by pchaigno — APPROVED** [review]

**Pull request merged**
